instrumentation/messaging/RATIONALE.md

```diff
@@ -78,3 +78,155 @@ continue incoming state.
+`SequenceId` field, which embeds parition information with a logical sequence. Note, when using
+Kafka protocol an `offset` will also exist, and is not directly related to the `SequenceId`.
+
+A Google PubSub Subscriber receives both a `message_id` and an `ack_id` in the `PullResponse`. The
+`ack_id` has a different format and is scoped to the subscriber.
+
+Pulsar has a client-generated `SequenceID`, but the broker controls the `MessageID` (sent in the
+response). The `MessageID` is not derived from the `SequenceID` and they serve different purposes.
+`SequenceID` is more about in-flight message tracking; consumer and admin apis use `MessageId` to
+identify, ack and nack a message.
+
+### Why use offset/sequence when there's no message ID?
+RocketMQ embeds offset in their message ID while Kafka and MQTT don't define a field named message
+ID. When there isn't a field named message ID, we use the closest stable value, such as a sequence
+number or offset, as opposed to returning `null` or synthesizing an ID from multiple fields.
+
+We use a sequence number or offset with the understanding that this could clarify duplicate sends or
+trace context breaks in the same way that a message ID could, even if it requires looking at other
+fields. Plus many message ID formats require looking at other fields anyway. A standard tag is
+easier to access as it requires no library specific types to parse.
+
+Ex. This requires just the messaging jar to express a policy that includes a message ID like field:
+```java
+MessagingTags.MESSAGE_ID.tag(req, context, span);
+```
+
+Ex. This requires messaging and Kafka instrumentation jars to tag the same:
+```java
+KafkaTags.OFFSET.tag(req, context, span);
+```
+
+Pragmatically, we choose to use a field that serves at least a common purpose as message ID where
+possible instead of returning `null` for lack of an exact match. Specific to Kafka, we choose the
+offset.
+
+### Why not synthesize a format that includes all needed fields when there's no message ID?
+Typical message ID formats encode multiple components such as a broker ID or network address,
+destination, timestamp or offset. It may be tempting to compose a format to include the dimensions
+that likely pinpoint a message when there's no format defined by the library. For example, in Kafka,
+we could compose a format like `topic-partition-offset` to ensure `MessagingRequest.id()` would have
+all identifying information.
+
+If we did that, we'd add overhead with the only consumer being tracing itself. It would fail as a
+correlation field with other libraries as by definition our format would be bespoke. Moreover,
+higher layers of abstraction which might have a defined message ID format could be confused with
+ours. Later, if that same tool creates a message ID format, it would likely be different from ours.
+
+For reasons including these, if there's no message ID field, and failing that just return `null`.
```

Comment:
:thinking:  ... ThisSentenceCannotBeGrokkedException 
```suggestion
For reasons including these, if there's no message ID field, and failing that just return `null`.
```
the `and failing that` part confuses me, what do you mean

Reply:
hopefully fixed